src/vm/mod.rs

```diff
@@ -45,6 +46,9 @@ where
     type VMActivePlan: ActivePlan<Self>;
     type VMReferenceGlue: ReferenceGlue<Self>;
 
+    /// The type of edges in this VM.
+    type VMEdge: edge_shape::Edge;
```

Comment:
Is it possible to do a runtime switch? For example, OpenJDK may need to support compressed pointers as an optional run-time feature, and cannot be activated at build-time.

Reply:
Also probably this can be moved to VMObjectModel?

Reply:
If the "feature" is a Cargo feature, we can define it like this:
```rust
enum OpenJDKEdge {
    Uncompressed(*mut AtomicUsize),
    #[cfg(features = "compressed_oop")]
    Compressed(*mut AtomicU32),
}
```

But if we want it enabled/disabled by a command-line flag, the easiest way (may affect performance, but need to be confirmed with tests) is simply not use the `OpenJDKEdge::Compressed` case.  As an optimisation, maybe we can parameterise `OpenJDK` to `OpenJDK<EdgeType>` so that depending on the command line options, we can instantiate MMTk with different (parameterised) VMBinding implementation.

Reply:
Maybe we could check the performance impact of having `enum Edge` (but never using the other case). We do hope that is minimal.

Reply:
> Also probably this can be moved to VMObjectModel?

I don't think `ObjectModel` is the right place to put it. I think it is fine to put the type in `VMBinding` (as it is now). If we plan to move it, I think `Scanning` fits well. All related changes are in the `Scanning` trait. This is an indication that the type is related with `Scanning`.

Reply:
> Also probably this can be moved to VMObjectModel?

In theory, we can.  But then instead of `VM::VMEdge`, we will need to write `<<VM as VMBinding>::VMObjectModel as ObjectModel<VM>>::VMEdge` everywhere.

A refactoring may be possible to let VMBinding *require* VMObjectModel, instead of including a VMObjectModel as a member.  I have seen similar problems when I tried to refactor ReferenceProcessor.  I tried to add a function in `VMCollection` that depends on a type in `VMReferenceGlue`, and I have to navigate from VMCollection to VMBinding, and from there to VMReferenceGlue.  Sometimes the compiler is not smart enough to figure out the VM type passed to the function to VMCollection is the same VM as the VM in `VMReferenceGlue<VM>`.

I'll elaborate the refactoring in a different issue.

Reply:
Using `enum OpenJDKEdge` will (1) introduce an extra branch for each edge load and store (2) For a mark-queue `Vec<Edge>`, there will be one word (100%) overhead for each entry. With compressed pointers on, the overhead will be 300%. Probably I'll parameterize `OpenJDK`.

Reply:
> In theory, we can. But then instead of VM::VMEdge, we will need to write <<VM as VMBinding>::VMObjectModel as ObjectModel<VM>>::VMEdge everywhere.

I see. Yes I agree it should stay in VMBinding.

Reply:
I think I don't really need another edge type for compressed pointers. The size of an edge **address** is always one word, no matter whether it's a `*mut AtomicU32` or `*mut AtomicUsize `. The only thing I need to replace is the handler of the load/store operations.

Reply:
If space overhead is a problem, the OpenJDK binding can implement `Edge` with a tagged pointer.  It saves space, but still needs a test on bits.

Other bindings may not be able to use this trick.  Ruby already uses tagged pointers, and Julia has offsetted edges where the offset has to be recorded.